tests/TestHarness/launcher.py

```diff
@@ -66,9 +188,264 @@ def comma_separated(string):
+
+    def define_network(self):
+        if self.args.per_host == 0:
+            for i in range(self.args.total_nodes):
+                node_name, cfg_name = self.assign_name(i == 0)
+                node = nodeDefinition(node_name, cfg_name, self.args.base_dir, self.args.config_dir, self.args.data_dir)
+                node.set_host(i == 0)
+                self.aliases.append(node.name)
+                self.network.nodes[node.name] = node
+        else:
+            ph_count = 0
+            host_ndx = 0
+            num_prod_addr = 1
+            num_nonprod_addr = 1
+            for i in range(self.args.total_nodes, 0, -1):
+                do_bios = False
+                node_name, cfg_name = self.assign_name(i == 0)
+                lhost = nodeDefinition(node_name, cfg_name, self.args.base_dir, self.args.config_dir, self.args.data_dir)
+                lhost.set_host(i == 0)
+                if ph_count == 0:
+                    if host_ndx < num_prod_addr:
+                        do_bios = True # servers.producer[host_ndx].has_bios
+                        lhost.host_name = '127.0.0.1' # servers.producer[host_ndx].ipaddr
+                        lhost.public_name = 'localhost' # servers.producer[host_ndx].name
+                        ph_count = 1 # servers.producer[host_ndx].instances
+                    elif host_ndx - num_prod_addr < num_nonprod_addr:
+                        ondx = host_ndx - num_prod_addr
+                        do_bios = True # servers.nonprod[ondx].has_bios
+                        lhost.host_name = '127.0.0.1' # servers.nonprod[ondx].ipaddr
+                        lhost.public_name = 'localhost' # servers.nonprod[ondx].name
+                        ph_count = 1 # servers.nonprod[ondx].instances
+                    else:
+                        lhost.host_name = f'pseudo_{host_ndx.zfil(2)}'
+                        lhost.public_name = lhost.host_name
+                        ph_count = 1
+                    host_ndx += 1
+                lhost.name = self.assign_name(do_bios)
+                self.aliases.append(lhost.name)
+                do_bios = False
+                ph_count -= 1
+
+    def bind_nodes(self):
+        if self.args.pnodes < 2:
+            raise RuntimeError(f'Unable to allocate producers due to insufficient pnodes = {self.args.pnodes}')
+        non_bios = self.args.pnodes - 1
+        per_node = self.args.producers / non_bios
+        extra = self.args.producers % non_bios
+        i = 0
+        producer_number = 0
+        to_not_start_node = self.args.total_nodes - self.args.unstarted_nodes - 1
+        for node_name, node in self.network.nodes.items():
+            is_bios = node_name == 'bios'
+            node.keys.append(KeyStrings('EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV',
+                                         '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3'))
+            if is_bios:
+                node.producers.append('eosio')
```

Comment:
lost ` producer_set.schedule.push_back({prodname,pubkey});`  Will keep eye open for how this may be handled elsewhere, but noting for now.

Reply:
Looks like this is part of the `setprods` format support for launcher and is used for `write_setprods_file` and for `write_bios_boot`, both of which are not yet ported.  My assumption is that, if `producer_set` is needed in the future it will come along for the write when `write_setprods_file` and/or `write_bios_boot` implementations are ported.